h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java

```diff
@@ -103,26 +148,53 @@ public void map(Chunk c, NewChunk nc) {
   private static class NumMatchTask extends MRTask<CatMatchTask> {
     double[] _values;
     double _noMatch;
-    NumMatchTask(double[] values, double noMatch) {
+    int _startIndex;
+    IcedHashMap<Double, Integer> _matchesIndexes;
+
+    NumMatchTask(double[] values, double noMatch, int startIndex) {
       _values = values;
       _noMatch = noMatch;
+      _startIndex = startIndex;
+      _matchesIndexes = new IcedHashMap<>();
     }
+ 
     @Override
     public void map(Chunk c, NewChunk nc) {
       int rows = c._len;
       for (int r = 0; r < rows; r++) {
-        double x = c.isNA(r) ? _noMatch : in(_values, c.atd(r), _noMatch);
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _values, c.atd(r), _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
-  private static double in(String[] matches, String s, double nomatch) {
-    return Arrays.binarySearch(matches, s) >= 0 ? 1 : nomatch;
+  private static double in(IcedHashMap<String, Integer> matchesIndexes, String[] matches, String s, double noMatch, int startIndex) {
+    Integer mapResult = matchesIndexes.get(s);
+    int match;
+    if (mapResult == null){
+      match = Arrays.binarySearch(matches, s);
```

Comment:
you removed the `sort` but you keep `binarySearch`: the latter works only if `matches` is sorted, and we don't want that, so I think you should switch to `indexOf`.
A corner case performance-wise would be something like
```python
incr = [str(i) for i in range(100)]
fr = H2OFrame(incr)
fr.match(incr)  # should return a col frame equivalent to incr
decr = incr.reverse()
fr.match(decr) # should return a col frame equivalent to decr
```

Reply:
You are correct—my mistake. We can use sort and binarySearch, but we must save mapping to the original order of match values. Or we can use indexOf, but for `double `values, we use the `binarySearchDoublesUlp` method, so indexOf won't work correctly for doubles.

Reply:
> We can use sort and binarySearch, but we must save mapping to the original order of match values

yeah, generating this mapping will work: for a while I thought it would be too costly to generate this, but I don't see why it would anymore, it's done in one single traversal…
We can then keep both the sorting and the binary search everywhere then, as soon as we had this mapping. Thanks @maurever